larpandoracontent/LArThreeDReco/LArTwoViewMatching/TwoViewThreeDKinkTool.cc

```diff
@@ -0,0 +1,479 @@
+
+float TwoViewThreeDKinkTool::GetXSamplingPoint(const CartesianVector &splitPosition1, const bool isForwardInX,
+    const TwoDSlidingFitResult &fitResult1, const TwoDSlidingFitResult &fitResult2) const
+{
+    // Nearest common x position
+    float xMin1(std::numeric_limits<float>::max()), xMax1(-std::numeric_limits<float>::max());
+    float xMin2(std::numeric_limits<float>::max()), xMax2(-std::numeric_limits<float>::max());
+    fitResult1.GetMinAndMaxX(xMin1, xMax1);
+    fitResult2.GetMinAndMaxX(xMin2, xMax2);
+
+    const float commonX(isForwardInX ? std::max(xMin1, xMin2) : std::min(xMax1, xMax2));
+
+    if (isForwardInX && ((commonX > xMax1) || (commonX > xMax2)))
+        throw StatusCodeException(STATUS_CODE_NOT_FOUND);
+
+    if (!isForwardInX && ((commonX < xMin1) || (commonX < xMin2)))
+        throw StatusCodeException(STATUS_CODE_NOT_FOUND);
+
+    // Layer step x position
+    float rL1(0.f), rT1(0.f);
+    fitResult1.GetLocalPosition(splitPosition1, rL1, rT1);
+    const int splitLayer(fitResult1.GetLayer(rL1));
+
+    const int lowLayer(std::max(fitResult1.GetMinLayer(), std::min(fitResult1.GetMaxLayer(), splitLayer - m_nLayersForKinkSearch)));
+    const int highLayer(std::max(fitResult1.GetMinLayer(), std::min(fitResult1.GetMaxLayer(), splitLayer + m_nLayersForKinkSearch)));
+
+    CartesianVector minus(0.f, 0.f, 0.f), plus(0.f, 0.f, 0.f);
+    PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, fitResult1.GetGlobalFitPosition(fitResult1.GetL(lowLayer), minus));
+    PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, fitResult1.GetGlobalFitPosition(fitResult1.GetL(highLayer), plus));
+
+    if (minus.GetX() > plus.GetX())
+    {
+        CartesianVector temporary(minus);
+        minus = plus;
+        plus = temporary;
+    }
+
+    const float layerStepX(isForwardInX ? plus.GetX() : minus.GetX());
+
+    // Final x position selection
+    const float chosenX(isForwardInX ? std::max(layerStepX, commonX) : std::min(layerStepX, commonX));
+    const float finalX(isForwardInX ? chosenX + m_additionalXStepForKinkSearch : chosenX - m_additionalXStepForKinkSearch);
+    return finalX;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool TwoViewThreeDKinkTool::IsALowestInX(const LArPointingCluster &pointingClusterA, const LArPointingCluster &pointingClusterB)
+{
+    if ((pointingClusterA.GetInnerVertex().GetPosition().GetX() < pointingClusterB.GetInnerVertex().GetPosition().GetX()) &&
+        (pointingClusterA.GetInnerVertex().GetPosition().GetX() < pointingClusterB.GetOuterVertex().GetPosition().GetX()))
+    {
+        return true;
+    }
+
+    if ((pointingClusterA.GetOuterVertex().GetPosition().GetX() < pointingClusterB.GetInnerVertex().GetPosition().GetX()) &&
+        (pointingClusterA.GetOuterVertex().GetPosition().GetX() < pointingClusterB.GetOuterVertex().GetPosition().GetX()))
+    {
+        return true;
+    }
```

Comment:
No strong view on this, but I wonder if this can be made a little more concise via:

```
const float xMinA{std::min(pointingClusterA.GetInnerVertex().GetPosition().GetX(), pointingClusterA.GetOuterVertex().GetPosition().GetX())};
const float xMinB{std::min(pointingClusterB.GetInnerVertex().GetPosition().GetX(), pointingClusterB.GetOuterVertex().GetPosition().GetX())};
return xMinA < xMinB;
```